lib/puppet/provider/cisco_portchannel_global/nxapi.rb

```diff
@@ -0,0 +1,256 @@
+  desc 'The new NXAPI provider.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  PC_GLOBAL_NON_BOOL_PROPS = [
+    :bundle_hash,
+    :bundle_select,
+    :hash_distribution,
+    :hash_poly,
+    :load_defer,
+    :rotate,
+  ]
+  PC_GLOBAL_BOOL_PROPS = [
+    :asymmetric,
+    :concatenation,
+    :resilient,
+    :symmetry,
+  ]
+  PC_GLOBAL_ALL_PROPS = PC_GLOBAL_NON_BOOL_PROPS + PC_GLOBAL_BOOL_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@pc_global',
+                                            PC_GLOBAL_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@pc_global',
+                                            PC_GLOBAL_BOOL_PROPS)
+
+  def initialize(value={})
+    super(value)
+    @pc_global = Cisco::PortChannelGlobal.globals[@property_hash[:name]]
+    @property_flush = {}
+    debug 'Created provider instance of cisco_portchannel_global'
+  end
+
+  def self.properties_get(global_id, v)
+    debug "Checking instance, global #{global_id}"
+    current_state = {
+      name: global_id
+    }
+
+    # Call node_utils getter for each property
+    PC_GLOBAL_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = v.send(prop)
+    end
+
+    PC_GLOBAL_BOOL_PROPS.each do |prop|
+      val = v.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    globals = []
+    Cisco::PortChannelGlobal.globals.each do |global_id, v|
+      globals << properties_get(global_id, v)
+    end
+    globals
+  end
+
+  def self.prefetch(resources)
+    globals = instances
+
+    resources.keys.each do |id|
+      provider = globals.find { |global| global.instance_name == id }
+      resources[id].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def instance_name
+    name
+  end
+
+  def properties_set
+    PC_GLOBAL_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      unless @property_flush[prop].nil?
+        @pc_global.send("#{prop}=", @property_flush[prop]) if
+          @pc_global.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    port_channel_load_balance_sym_concat_rot_set
+    port_channel_load_balance_hash_poly_set
+    port_channel_load_balance_asym_rot_set
+  end
+
+  # for n9k/n3k:
+  # port-channel load-balance src-dst ip rotate 4 concatenation symmetric
+  # all the above properties will be set all at once so make sure
+  # all the needed resources are present
+  def port_channel_load_balance_sym_concat_rot_all?
+    @resource[:bundle_hash] &&
+      @resource[:bundle_select] &&
+      @resource[:concatenation] &&
+      @resource[:symmetry] &&
+      @resource[:rotate]
+  end
+
+  def port_channel_load_balance_sym_concat_rot_set
```

Comment:
beaker tests themselves are ok but I have to try the demo manifest because that is where when things are a little bit more stressed when device side properties and manifest params are same. I have to try this out first and will let you know what I find.